osquery/sql/sqlite_collation.cpp

```diff
@@ -0,0 +1,257 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <string>
+
+#include <sqlite3.h>
+
+namespace osquery {
+/**
+ * @brief Compares the position of epoch delimiter between two versions.
+ * Return difference of epoch segment length if both versions have epoch.
+ * Return 1 if left version has epoch but not the right.
+ * Return -1 if right version has epoch but not the left.
+ * Return 0 if epoch doesn't exist in either version.
+ */
+static int compareEpoch(int lLen,
+                        const char* lVer,
+                        int rLen,
+                        const char* rVer) {
+  auto lEpoch = strcspn(lVer, ":");
+  auto rEpoch = strcspn(rVer, ":");
+
+  if (lEpoch != lLen && rEpoch != rLen) {
+    return lEpoch - rEpoch;
+  } else if (lEpoch != lLen && rEpoch == rLen) {
+    return 1;
+  } else if (lEpoch == lLen && rEpoch != rLen) {
+    return -1;
+  }
+
+  return 0;
+}
+
+/**
+ * @brief Check if character is defined as a delimiter: `~-^.:`
+ */
+static bool isDelimiter(int c) {
```

Comment:
Does this need to be separate from `delimiterPrecedence`? Is there a performance win?

Reply:
I'll clean that up. Not really any difference worth having two of them for.